File: lecture/inflearn-spring-boot/autoconfig/src/main/java/hello/repository/MemberRepository.java
package hello.repository;

import hello.member.Member;
import java.util.List;
import org.springframework.jdbc.core.BeanPropertyRowMapper;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.stereotype.Repository;

@Repository
public class MemberRepository {

  private final JdbcTemplate jdbcTemplate;

  public MemberRepository(JdbcTemplate jdbcTemplate) {
    this.jdbcTemplate = jdbcTemplate;
  }

  public void initTable() {
    jdbcTemplate.execute(
      "create table member(member_id varchar primary key, name varchar)"
    );
  }

  public void save(Member member) {
    jdbcTemplate.update("insert into member(member_id, name) values(?, ?)", member.getMemberId(), member.getName());
  }

  public Member find(String memberId) {
    return jdbcTemplate.queryForObject(
      "select member_id, name from member where member_id=?",
      BeanPropertyRowMapper.newInstance(Member.class),
      memberId
    );
  }

  public List<Member> findAll() {
    return jdbcTemplate.query(
      "select member_id, name from member",
      BeanPropertyRowMapper.newInstance(Member.class)
    );
  }
}

File: lecture/fs-mysql/fs-mysql/src/main/kotlin/com/lecture/fsmysql/domain/member/repository/MemberRepository.kt
package com.lecture.fsmysql.domain.member.repository

import com.lecture.fsmysql.domain.member.entity.Member
import org.springframework.jdbc.core.RowMapper
import org.springframework.jdbc.core.namedparam.BeanPropertySqlParameterSource
import org.springframework.jdbc.core.namedparam.MapSqlParameterSource
import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate
import org.springframework.jdbc.core.simple.SimpleJdbcInsert
import org.springframework.stereotype.Repository
import java.time.LocalDate
import java.time.LocalDateTime

@Repository
class MemberRepository(
  private val namedParameterJdbcTemplate: NamedParameterJdbcTemplate
) {

  companion object {
    private const val TABLE = "Member"
  }

  fun save(member: Member): Member {
    return member.id?.let {
      update(member)
    } ?: insert(member)
  }

  private fun insert(member: Member): Member {
    val simpleJdbcInsert = SimpleJdbcInsert(namedParameterJdbcTemplate.jdbcTemplate)
      .withTableName(TABLE)
      .usingGeneratedKeyColumns("id")
    val parameter = BeanPropertySqlParameterSource(member)
    val id = simpleJdbcInsert.executeAndReturnKey(parameter).toLong()
    return Member(
      id = id,
      nickname = member.nickname,
      email = member.email,
      birthday = member.birthday,
    )
  }

  private fun update(member: Member): Member {

    return member
  }

  fun findById(id: Long): Member? {
    val sql = String.format("select * from %s where id = :id", TABLE)
    val parameter = MapSqlParameterSource()
      .addValue("id", id)
    val rowMapper = RowMapper { resultSet, rowNum ->
      Member(
        id = resultSet.getLong("id"),
        nickname = resultSet.getString("nickname"),
        email = resultSet.getString("email"),
        birthday = resultSet.getObject("birthday", LocalDate::class.java),
        createdAt = resultSet.getObject("createdAt", LocalDateTime::class.java)
      )
    }
    return namedParameterJdbcTemplate.queryForObject(sql, parameter, rowMapper)
  }
}